pdftk: New package

citibeth opened the pull request:
pdftk is a nifty command-line utility to pull apart and rearrange PDFs.  And one I rely on, so that's why I'm putting it into Spack.

This is a unique package (from Spack's point of view) because it is written in a combination of Java and C++, and REQUIRES GCJ to build.  This package works, but I'm not sure how to best do a lot of the unique things it needs to do.  I'm looking for feedback.

Notes:

1. It requires a compiler other than C/C++/Fortran/Fortran77.  So I'm not sure how this would best fit into Spack's compiler infrastructure.

2. It ONLY builds with GCC.  There's no compatiblity with Intel or anything else, because GCC is the only compiler that provides GCJ.

3. GCJ has been removed from GCC as of version 7, so this requires an "older" version of GCC.

4. There are some problems getting GCJ working properly (see #8165).  Those issues could be addressed in the `gcc` package, or in an auxiliary package that would be needed for anything requiring `gcj` --- this package is likely the only thing that will ever need that.

citibeth commented:
@adamjstewart Can I get anyone to look at this package?  I know it's unusual...

adamjstewart commented on var/spack/repos/builtin/packages/pdftk/package.py:
```diff
@@ -0,0 +1,110 @@
+
+
+class Pdftk(MakefilePackage):
+    """Parallel Ice Sheet Model"""
```
Is this the docstring for `pdftk` or for `pism`?

adamjstewart commented on var/spack/repos/builtin/packages/pdftk/package.py:
```diff
@@ -0,0 +1,110 @@
+    def build(self, spec, prefix):
+        compiler = self.compiler
+        gcc_base = os.path.split(os.path.split(compiler.cxx)[0])[0]
+        env = dict(os.environ)
```
This is already available in the environment, no need to create it again.

adamjstewart commented on var/spack/repos/builtin/packages/pdftk/package.py:
```diff
@@ -0,0 +1,110 @@
+        gcc_base = os.path.split(os.path.split(compiler.cxx)[0])[0]
+        env = dict(os.environ)
+        env['PATH'] = join_path(gcc_base, 'bin') + ':' + env['PATH']
+        os.chdir(self.build_directory)
```
It's preferable to use `with working_dir(self.build_directory):`

adamjstewart commented on var/spack/repos/builtin/packages/pdftk/package.py:
```diff
@@ -0,0 +1,110 @@
+#        try:
+#            make('-f', 'Makefile.Spack')
+#        except:
+#            pass
```
If you aren't using these lines can you remove them from the package?

adamjstewart commented on var/spack/repos/builtin/packages/pdftk/package.py:
```diff
@@ -0,0 +1,110 @@
+#            pass
+
+        # Finish the build in serial
+        make('-f', 'Makefile.Spack', '-j1')
```
Spack already adds `-j{# procs}` to the `make` call by default. If you need it to build in serial, add `parallel=False`.

adamjstewart commented on var/spack/repos/builtin/packages/pdftk/package.py:
```diff
@@ -0,0 +1,110 @@
+        make('-f', 'Makefile.Spack', '-j1')
+
+    def install(self, spec, prefix):
+        # "make install" is hardcoded to /usr...
```
You could filter the makefile. That might be an easier option than all of this.

P.S. These functions are already available, no need to call the full `llnl.util.filesystem` path.

adamjstewart commented on var/spack/repos/builtin/packages/pdftk/package.py:
```diff
@@ -0,0 +1,110 @@
+#        # This is only needed to solve the ej1 stuff see #8165
+#        compiler = self.compiler
+#        gcc_base = os.path.split(os.path.split(compiler.cxx)[0])[0]
+#        spack_env.prepend_path('PATH', join_path(gcc_base, 'bin'))
```
Remove if not using.

citibeth commented:
I believe that GCC is the only compiler collection that ever offered Java
-> binary translation.  So yes, it only builds with GCC.

On Fri, May 18, 2018 at 4:53 PM, Adam J. Stewart <notifications@github.com>
wrote:

> *@adamjstewart* requested changes on this pull request.
>
> You're right that this package is... weird. Does it only build with GCC?
> ------------------------------
>
> In var/spack/repos/builtin/packages/pdftk/package.py:
>
> > +# This program is distributed in the hope that it will be useful, but
> +# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
> +# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
> +# conditions of the GNU Lesser General Public License for more details.
> +#
> +# You should have received a copy of the GNU Lesser General Public
> +# License along with this program; if not, write to the Free Software
> +# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111 <https://maps.google.com/?q=59+Temple+Place,+Suite+330,+Boston,+MA+02111&entry=gmail&source=g>-1307 USA
> +##############################################################################
> +from spack import *
> +import os
> +import llnl.util.filesystem
> +
> +
> +class Pdftk(MakefilePackage):
> +    """Parallel Ice Sheet Model"""
>
> Is this the docstring for pdftk or for pism?
> ------------------------------
>
> In var/spack/repos/builtin/packages/pdftk/package.py:
>
> > +            ('RMFLAGS', '-vf'),
> +            ('CPPFLAGS', ' '.join(cppflags)),
> +            ('CXXFLAGS', ' '.join(cxxflags)),
> +            ('GCJFLAGS', ' '.join(gcjflags)),
> +            ('GCJHFLAGS', '-force'),
> +            ('LDLIBS', '-lgcj')
> +        ]
> +        with open(join_path('pdftk', 'Makefile.Spack'), 'w') as mk:
> +            for var, val in vars:
> +                mk.write("export {0}={1}\n".format(var, str(val)))
> +            mk.write('include Makefile.Base\n')
> +
> +    def build(self, spec, prefix):
> +        compiler = self.compiler
> +        gcc_base = os.path.split(os.path.split(compiler.cxx)[0])[0]
> +        env = dict(os.environ)
>
> This is already available in the environment, no need to create it again.
> ------------------------------
>
> In var/spack/repos/builtin/packages/pdftk/package.py:
>
> > +            ('CXXFLAGS', ' '.join(cxxflags)),
> +            ('GCJFLAGS', ' '.join(gcjflags)),
> +            ('GCJHFLAGS', '-force'),
> +            ('LDLIBS', '-lgcj')
> +        ]
> +        with open(join_path('pdftk', 'Makefile.Spack'), 'w') as mk:
> +            for var, val in vars:
> +                mk.write("export {0}={1}\n".format(var, str(val)))
> +            mk.write('include Makefile.Base\n')
> +
> +    def build(self, spec, prefix):
> +        compiler = self.compiler
> +        gcc_base = os.path.split(os.path.split(compiler.cxx)[0])[0]
> +        env = dict(os.environ)
> +        env['PATH'] = join_path(gcc_base, 'bin') + ':' + env['PATH']
> +        os.chdir(self.build_directory)
>
> It's preferable to use with working_dir(self.build_directory):
> ------------------------------
>
> In var/spack/repos/builtin/packages/pdftk/package.py:
>
> > +
> +    def build(self, spec, prefix):
> +        compiler = self.compiler
> +        gcc_base = os.path.split(os.path.split(compiler.cxx)[0])[0]
> +        env = dict(os.environ)
> +        env['PATH'] = join_path(gcc_base, 'bin') + ':' + env['PATH']
> +        os.chdir(self.build_directory)
> +        print('build', os.getcwd())
> +
> +        # Parallel build is known to fail.  But let it get most
> +        # of the way there.
> +        # (Not sure if this actually saves any time)
> +#        try:
> +#            make('-f', 'Makefile.Spack')
> +#        except:
> +#            pass
>
> If you aren't using these lines can you remove them from the package?
> ------------------------------
>
> In var/spack/repos/builtin/packages/pdftk/package.py:
>
> > +        gcc_base = os.path.split(os.path.split(compiler.cxx)[0])[0]
> +        env = dict(os.environ)
> +        env['PATH'] = join_path(gcc_base, 'bin') + ':' + env['PATH']
> +        os.chdir(self.build_directory)
> +        print('build', os.getcwd())
> +
> +        # Parallel build is known to fail.  But let it get most
> +        # of the way there.
> +        # (Not sure if this actually saves any time)
> +#        try:
> +#            make('-f', 'Makefile.Spack')
> +#        except:
> +#            pass
> +
> +        # Finish the build in serial
> +        make('-f', 'Makefile.Spack', '-j1')
>
> Spack already adds -j{# procs} to the make call by default. If you need
> it to build in serial, add parallel=False.
> ------------------------------
>
> In var/spack/repos/builtin/packages/pdftk/package.py:
>
> > +        os.chdir(self.build_directory)
> +        print('build', os.getcwd())
> +
> +        # Parallel build is known to fail.  But let it get most
> +        # of the way there.
> +        # (Not sure if this actually saves any time)
> +#        try:
> +#            make('-f', 'Makefile.Spack')
> +#        except:
> +#            pass
> +
> +        # Finish the build in serial
> +        make('-f', 'Makefile.Spack', '-j1')
> +
> +    def install(self, spec, prefix):
> +        # "make install" is hardcoded to /usr...
>
> You could filter the makefile. That might be an easier option than all of
> this.
>
> P.S. These functions are already available, no need to call the full
> llnl.util.filesystem path.
> ------------------------------
>
> In var/spack/repos/builtin/packages/pdftk/package.py:
>
> > +
> +        # Finish the build in serial
> +        make('-f', 'Makefile.Spack', '-j1')
> +
> +    def install(self, spec, prefix):
> +        # "make install" is hardcoded to /usr...
> +        llnl.util.filesystem.mkdirp(self.spec.prefix.bin)
> +        llnl.util.filesystem.install(
> +            join_path(self.stage.source_path, 'pdftk', 'pdftk'),
> +            self.spec.prefix.bin)
> +
> +#    def setup_environment(self, spack_env, run_env):
> +#        # This is only needed to solve the ej1 stuff see #8165
> +#        compiler = self.compiler
> +#        gcc_base = os.path.split(os.path.split(compiler.cxx)[0])[0]
> +#        spack_env.prepend_path('PATH', join_path(gcc_base, 'bin'))
>
> Remove if not using.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/8166#pullrequestreview-121568355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cdwBBM4h7df-cPkSdreODR5ap5LTmks5tzzTUgaJpZM4UCUxe>
> .
>

adamjstewart commented:
@citibeth I'm not sure if I'm qualified to answer that question. I completely agree that Spack works great if you happen to need a C/C++/Fortran compiler but grinds to a halt as soon as you need another compiler language. I was hoping someone would reply to #8187 as it is a great question.

citibeth commented:
The unit tests failed with a strange non-error https://travis-ci.org/spack/spack/jobs/381780767

```
...
lib/spack/spack/test/config.py .................
lib/spack/spack/test/database.py .................
lib/spack/spack/test/directory_layout.py ....
lib/spack/spack/test/environment.py ..........
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```
I'll try re-running...

citibeth commented:
> Still uncomfortable about the Java stuff. I want to see other people comment on this PR before merging.

@scheibelp 

We should think of `pdftk` not as being in Java, but rather its own special language I will call "GCJ".  It's like Java, but not the same --- and completely different in the way it's built.  GCC is the only compiler for that "language".

The GCJ compiler can be thought of as being able to do two things:
 1. Compile from Java bytecode to machine binary
 2. Compile from Java source code to machine binary.

It can do (1) out of the box.  But it doesn't actually come with a GNU-supplied parser, so it can't do (2).  To do that, it uses the Eclipse Java Parser (ecj), which is a separate download.  The ecj is written in Java, of course and is normally expected to run in a Java Virtual Machine (JVM), not compiled as binaries.  But in GCC land, there is no JVM.  So the ecj must be compiled to binary.  This is done using gcj in mode (1).

The upshot is... to compile Java in Spack, you need to `depends_on('eclipse-gcj-parser')`.  This package then verifies that it's actually using GCC (and not Intel or whatever); and it provides you the `gcj` executable, via: `spec['eclipse-gcj-parser'].package.gcj`.  Compiler wrappers are not needed because `gcj` is just used to build `.o` files, and not link; so no issues with RPATHs.

In general when we have "non-C/F" compilers, we depend on them as a build dependency.  That model doesn't quite work here because the required compiler is actually GCC; just a little-used part of it.  However, `eclipse-gcj-parser` serves a similar purpose.

The alternative to this would be to promote `gcj` to a "first class" compiler.  I think that would cause a lot of hassle for the 99% of Spack users who will never use GCJ.  Java is not really "a thing" in HPC, and GCJ has been deprecated since GCC7.  But `pdftk` is really useful --- useful enough to continue building it, even if it requires an old compiler.  I expect that `pdftk` will be the ONLY Spack package that needs this stuff.

Maybe someday someone will convert `pdftk` to C++; or make it work with a "regular" Java system.

citibeth commented:
@adamjstewart How do I add a compiler dependency?  I want the equivalent of `depends_on(gcc@:6)`.

On a philosophical note: I mentioned:
```
`pdftk` is clearly "on the way out" of mainstream Linux distros.  It left Fedora a number of years ago; and the last Ubuntu to include it is 17.10.  It DOES continue to exist in Debian for the time being.

Removal from Linux distros is why it's important that Spack can build this.  AFAIK, there are alternatives with similar functionality.
```

Spack is better suited to support `pdftk` than a typical Linux distro, due to its multi-version nature.  It may be a PITA to install an old GCC just to build `pdftk`.  But it's possible, and it doesn't "mess up" the rest of the system.  Linux distros are forced to drop `pdftk` as they move to GCC7, which no longer supports GCJ.

adamjstewart commented:
> How do I add a compiler dependency? I want the equivalent of `depends_on(gcc@:6)`.

There is currently no way to do that, although @tgamblin and I have discussed making compilers closer to dependencies for a long time. Ideally, I would like to see something like:
```python
requires('%gcc@:6')
```
or
```python
requires('c++11')
```
The best you can do right now is add `conflicts()` for every other compiler.

citibeth commented:
I think I'll leave it as-is then.  It does check for GCC at runtime, and
will fail appropriately.

On Tue, May 22, 2018 at 11:41 AM, Adam J. Stewart <notifications@github.com>
wrote:

> How do I add a compiler dependency? I want the equivalent of
> depends_on(gcc@:6).
>
> There is currently no way to do that, although @tgamblin
> <https://github.com/tgamblin> and I have discussed making compilers
> closer to dependencies for a long time. Ideally, I would like to see
> something like:
>
> requires('%gcc@:6')
>
> or
>
> requires('c++11')
>
> The best you can do right now is add conflicts() for every other compiler.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/8166#issuecomment-391039003>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cdzaSct5EoVI9rW4YM2yRL774C9yGks5t1DGbgaJpZM4UCUxe>
> .
>

scheibelp commented:
This failed flake checks. If that gets addressed by 6/11 I should be able to merge it then (based on skimming this I don't immediately see any major issues)

imnasnainaec commented:
It fails for me:

$ spack install pdftk
==> Installing eclipse-gcj-parser \\
==> Using cached archive: /home/dror/spack/var/spack/cache/eclipse-gcj-parser/eclipse-gcj-parser-4.8.None
==> Already staged eclipse-gcj-parser-4.8-2bmbm5sqmedhow6pgjuee2fk7rcrg56o in /home/dror/spack/var/spack/stage/eclipse-gcj-parser-4.8-2bmbm5sqmedhow6pgjuee2fk7rcrg56o
==> No patches needed for eclipse-gcj-parser
==> Building eclipse-gcj-parser [Package]
==> Executing phase: 'build'
==> Error: ProcessError: /usr/bin/gcj: No such file or directory
    Command: '/usr/bin/gcj' '-o' 'ecj1' '--main=org.eclipse.jdt.internal.compiler.batch.GCCMain' 'ecj-4.8.jar'
See build log for details:
  /home/dror/spack/var/spack/stage/eclipse-gcj-parser-4.8-2bmbm5sqmedhow6pgjuee2fk7rcrg56o/spack-build.out
